icu4j/main/classes/core/src/com/ibm/icu/text/UnicodeSet.java

```diff
@@ -3930,97 +3961,185 @@ private static final void addCaseMapping(UnicodeSet set, int result, StringBuild
         // see UCaseProps
     }
 
+    /** For case closure on a large set, look only at code points with relevant properties. */
+    UnicodeSet maybeOnlyCaseSensitive(UnicodeSet src) {
+        if (src.size() < 30) {
+            return src;
+        }
+        return CharacterProperties.getBinaryPropertySet(UProperty.CASE_SENSITIVE).
```

Comment:
Not a blocker.
Would it be faster to clone the src and then retainAll CharacterProperties.getBinaryPropertySet(UProperty.CASE_SENSITIVE), because the src is likely to be smaller? Or maybe clone the smaller of the two?

Reply:
Hard to say without a benchmark. It's probably cheaper to clone the one with fewer *ranges*, and if the src.hasStrings() then cloning src unnecessarily clones the structure for that too.

Do you want me to do something like
```
UnicodeSet sensitive = ...
if (src.hasStrings() || src.getRangeCount() > sensitive.getRangeCount()) {
    return sensitive.cloneAsThawed().retainAll(src);
} else {
    return src.clone().retainAll(sensitive);
}
```
?

Reply:
You have a good point. Probably not worth doing without a benchmark.

Reply:
I just did this in the latest commit.